### PR TITLE
Add option to use a custom amount decimal format

### PR DESCRIPTION
**Pull request by ebellani**

### CODE REVIEW

**Comment by tomszilagyi** [comment]
Can you please motivate this, i.e. what problem are you trying to solve? I
do not see any newly added test case that was not correctly handled before.
Also, it would be good to keep the garbage-resilience, i.e. keep re-find as
the first step in convert-amount.

2018-04-01 23:29 GMT+02:00 Eduardo Bellani <notifications@github.com>:

**Comment by ebellani** [comment]
@tomszilagyi, the case in particular is for a number like `10,123.45`. In general, it is for numbers that are on different locales.

Removing the initial regexp cleaning was motivated by the cleaning conflicting with what the user might configure as a number pattern.

**Comment by ebellani** [comment]
@tomszilagyi, you were right, it was a bad move to edit those cases away in the unit test. Undoing so uncovered severe limitations in the approach I had used.

I re-added those cases and others, alongside commentaries. Hopefully this maintains the integrity of the current system while allowing for more options in dealing with amounts.

Let me know what you think, thanks!

**Comment by tomszilagyi** [comment]
Presenting so many options to the user is unnecessary, and to be honest,
confusing. I doubt we need to be so general as to allow setting these
separators to any value. In real life, either one of them is . and the
other , or it's the other way around. I would like to see an approach that
assumes a sensible regex match (containing an optional leading minus
followed by a series of digits interspersed with ./, and maybe space as
thousands separator) to be a number, discard any garbage before and after,
and convert that in a robust fashion, signalling failure instead of
silently returning a wrong result in case of doubt.

**Comment by ebellani** [comment]
@tomszilagyi, I understand that you feel a lot of options can be confusing to the end user, which could lead to them not using the system. My approach in regards to this was to set the defaults to, hopefully, sane defaults.

About what we see in real life, the I'd say that the variability can be [surprising](https://en.wikipedia.org/wiki/Decimal_separator#Examples_of_use), to say the least. It seems to me that regular expressions are very hard to get right for cases that have this amount of variability.

What do you think?

File: README.md [hunk 1]
```diff
@@ -45,6 +45,25 @@ run`.
              default: 2
         -t : Text (descriptor) column index specs (zero-based)
              default: %3
+       -af : A string that sets the decimal format used for the
+          amount. This should be used in combination with the `ds` and `gs`
+          options to have control over the format of the amount. Note:
+          Fixing a prefix and a sufix in this pattern is not needed, since
```
Comment (tomszilagyi) [inline README.md]: suffix

File: src/banks2ledger/core.clj [hunk 2]
```diff
@@ -178,7 +178,31 @@
...
    {:opt "-t" :value "%3"
-    :help "Text (descriptor) column index specs (zero-based)"}})
+    :help "Text (descriptor) column index specs (zero-based)"}
+
+   :amount-format
+   {:opt "-af" :value "#"
+    :help "A string that sets the decimal format used for the
+    amount. This should be used in combination with the `ds` and `gs`
+    options to have control over the format of the amount. Note:
+    Fixing a prefix and a sufix in this pattern is not needed, since
```
Comment (tomszilagyi) [inline src/banks2ledger/core.clj]: suffix

File: src/banks2ledger/core.clj [hunk 3]
```diff
@@ -233,16 +257,29 @@
    (.parse (java.text.SimpleDateFormat. (get-arg args-spec :date-format))
            datestr)))
 
+;; Removes everything up to a digit (with an optional minus char) in a
+;; string. The purpose of this is to enable the `convert-amount`
+;; function to deal with unexpected prefixes in amount values.
+(defn remove-up-to-a-digit [s]
```
Comment (tomszilagyi) [inline src/banks2ledger/core.clj]: please rename to "remove-leading-garbage"

File: test.sh [hunk 4]
```diff
@@ -26,7 +26,7 @@ fi
 
 echo -n "testing ica.csv... "
 lein run -l $LEDGER -f test/data/ica.csv -F ';' -sa 1 -m 4 -t '%1' \
-	-a 'Assets:ICA Account' > test/data/ica.out
+	-a 'Assets:ICA Account' -af "# kr" -ds "," -gs "." > test/data/ica.out
```
Comment (tomszilagyi) [inline test.sh]: You can't supply "# kr" here as an option, the point of garbage being that the user does not know (or care) about what it is before receiving the data. Conversion must work without requiring such precise a-priori specs given by the user.

**Comment by ebellani** [comment]
@tomszilagyi, what is your opinion now?

**Comment by tomszilagyi** [comment]
Will check when I have time

**Comment by tomszilagyi** [comment]
Here's the next one for you :)
How do I parse numbers such as "5 000,00" ?
I actually have these in one of the files, and I cannot parse them with this PR.
I tried `-ds ',' -gs ' '` and even `-af "# #,#"` but that does not work.

**Comment by ebellani** [comment]
@tomszilagyi, the added example may cover what you are looking for.

**Comment by tomszilagyi** [comment]
Okay, thanks, this looks good. I like it.

I have one final question/concern regarding the `-af` option. I have experimented a bit and found that all tests pass (even those that I have added myself and not yet published) if I set the `amount-format` to `###,###.#` (for those that were not already set to something similar, i.e., were at the default `#`). This makes life much simpler for the user -- I, for one, had no idea that `-ds ',' -gs ' '` alone is not sufficient for `"5 000,00"` to parse correctly, and while I now see that it is technically possible with the "correct" `-af`, it still does not make sense to me (as a user).

So... is there anything against setting `###,###.#` as the default for `amount-format`? The benefit would be that one does not have to specify `-af` for any of the problematic cases we covered so far (i.e., using only `-ds ',' -gs ' '` will now work on `"5 000,00"` and all similar cases will give the correct result). This would be step 1. But I would even go further and suggest that with this change to the default, the `-af` option should be removed entirely (step 2), unless you can give me an example where it is necessary.

Let me know if you think step 1 alone or maybe even with step 2 is a good idea.

**Comment by ebellani** [comment]
@tomszilagyi, I think option number 2 is a fantastic suggestion.

I've modified the code to incorporate it, let me know what you think.

**Comment by tomszilagyi** [comment]
Wonderful, I will wrap this up when I get a bit of time.

**Pull request merged**